Common: Consume, proxy and insert metadata from downstream to upstream

soya3129 opened the pull request:
*Description*: The change enables request metadata consuming, proxying and inserting. Detailed implementation is described in the doc source/docs/h2_metadata.md. 

*Risk Level*: Low. Not run in production.
*Testing*: Integration test.
*Docs Changes*: source/docs/h2_metadata.md
#2394

soya3129 commented:
Cc birenroy@.

soya3129 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -893,6 +906,13 @@ void ConnectionManagerImpl::ActiveStream::decodeData(ActiveStreamDecoderFilter*
 
+    if (!request_metadata_map_vector_.empty()) {
+      for (const auto& metadata_map : request_metadata_map_vector_) {
+        decodeMetadata((*entry).get(), *metadata_map);
```
In the request direction, it's more straight-forward to pass the metadata to downstream filters only. But to be consistent with the response direction, shall I pass metadata through all filters? I think we may be able to do something similar on the response direction to enable downstream only filter pass. What do you think?

soya3129 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -782,6 +782,19 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
+      // indicate the end of the stream.
+      if ((*entry)->end_stream_) {
+        Buffer::OwnedImpl empty_data("");
+        addDecodedData(*((*entry).get()), empty_data, true);
```
Apologize for the long response!

Added a trace log.

Great catch!! It wasn't working with decoding_headers_only_! To solve this problem, I tried to pass the empty data frame (added by addDecodedData) through filters when decoding_headers_only_ is true, but failed. The main problem is hard to distinguish this empty data frame with the data frames that are received from downstream (which we will ignore because of decoding_headers_only_). We could add more conditions to allow empty data frame to go through the filters here: https://github.com/envoyproxy/envoy/blob/208a7fc72dd5f63fe1722111a248de4ce9224eab/source/common/http/conn_manager_impl.cc#L822. But then I can't think of a way to set end_stream true in the empty data frame. I think complete() has to be true in order to set the end_stream in the empty data frame (https://github.com/envoyproxy/envoy/blob/208a7fc72dd5f63fe1722111a248de4ce9224eab/source/common/http/conn_manager_impl.cc#L1529). However, if I set complete() to be true, the newly received data frame from downstream will cause assert failure (complete() is true means we are done with downstream, shouldn't receive more data).

A very hacky solution I used is to add an empty data in router filter when end_stream in headers is true and we still have metadata to send. The downside is the data frame won't go through filters (Since filters decides to process headers only, it might be OK). The upside is it's simple.

But now the code is messy: if metadata is added in headers only request, the added empty data frame will go through filters. If metadata is added with decoding_headers_only_ true, the added empty data will not go through filters. I could merge those two cases to one: only add new empty data in router filters. What do you think?

About test: GrpcHttp1ReverseBridge returns ContinueAndEndStream in encodeHeaders(). So I used decode-headers-only to test, which returns ContinueAndEndStream in decodeHeaders(). Did I miss anything?

Please let me know any suggestions/thoughts. Thanks!!

soya3129 replied:
Metadata is sent as h2 extension frames. Some implementations may ignore or mishandle this frame type. It could potentially cause problems if we allow metadata to carry end_stream. But if we think it is not a concern, I can change to allow end_stream. If we do so, we may differ from GFE implementation.

soya3129 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -893,6 +906,13 @@ void ConnectionManagerImpl::ActiveStream::decodeData(ActiveStreamDecoderFilter*
     ENVOY_STREAM_LOG(trace, "decode data called: filter={} status={}", *this,
                      static_cast<const void*>((*entry).get()), static_cast<uint64_t>(status));
 
+    if (!request_metadata_map_vector_.empty()) {
```
Added a utility function. Thanks!

soya3129 commented on source/common/http/conn_manager_impl.h:
```diff
@@ -423,6 +426,7 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     absl::optional<Router::RouteConstSharedPtr> cached_route_;
     absl::optional<Upstream::ClusterInfoConstSharedPtr> cached_cluster_info_;
     DownstreamWatermarkCallbacks* watermark_callbacks_{nullptr};
+    MetadataMapVector request_metadata_map_vector_;
```
Fixed. Thanks!

mattklein123 replied:
2 things for thought:
1) Related to some of my other comments, can we potentially just be directly storing pointers to metadata maps so that we can avoid copies?
2) I'm wondering if we want to actually have some type of metadata storage structure that is created on demand? So few people will be using this feature that I'm wondering if it's worth the memory and small init cost in the common case? Thoughts?

soya3129 replied:
To answer 1: Storing metadata as unique_ptr<MetadataMap> instead of MetadataMapVector (aka std::vector<unique_ptr<MetadataMap>>) will merge two metadata with the same key value into one. This may cause some problem when multiple copies of the same metadata are expected, for example, for logging purpose.

To answer 2: I think that's good suggestion. Do you mind I make that change in a followup PR? Added a TODO.

mattklein123 replied:
> To answer 2: I think that's good suggestion. Do you mind I make that change in a followup PR? Added a TODO.

I think that's OK, but will you do it relatively quickly? I think we need to be mindful of the performance/memory impact of rarely used code in the hottest paths.

soya3129 replied:
I see. Fixed in this PR.

soya3129 commented on source/common/router/router.cc:
```diff
@@ -448,6 +448,22 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::HeaderMap& trailers) {
+  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata);
+
+  if (upstream_request_ == nullptr) {
+    ENVOY_STREAM_LOG(trace, "upstream_request_ not ready. Store metadata_map to encode later: {}",
```
Fixed. Thanks!

soya3129 commented on source/common/router/router.cc:
```diff
@@ -1019,6 +1035,14 @@ void Filter::UpstreamRequest::encodeData(Buffer::Instance& data, bool end_stream
+      ENVOY_STREAM_LOG(trace, "Send metadata before sending data. {}", *parent_.callbacks_,
+                       parent_.downstream_metadata_map_vector_);
+      request_encoder_->encodeMetadata(parent_.downstream_metadata_map_vector_);
+      parent_.downstream_metadata_map_vector_.clear();
```
Great point!! Thanks! I didn't think of those cases at all.

I am still working on this part. But I agree we should have a design for them before submitting this PR in case there are changes. Will update this thread when I make progress.

soya3129 replied:
I only have looked at the internal redirect case. Since those cases all call recreateStream(), I think they can be solved similarly.

recreateStream() calls decodeHeaders() to go through all filters at the end of the the function. https://github.com/envoyproxy/envoy/blob/4f4280ccd00b0e3b057f4c82b13bf6c4513c4ef2/source/common/http/conn_manager_impl.cc#L1789 . This allows us to add metadata again for the redirected request, so I think we may not need extra handling for the redirect case in envoy.

The issue is in nghttp2. In the redirect case, metadata ends up in session->ob_reg: https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2352, and the redirect headers ends up in session->ob_syn: https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2360 . Because session->ob_reg is parsed before session->ob_syn, metadata is received before headers (no new stream has created), and caused failure.

I switched the positions of the pointed code, and I don't see new test failures in nghttp2. And a simple internal redirect test case passed with metadata in envoy. (cl/231599606. If others are interested in the change, I can make a public available PR).

What do you think about making the nghttp2 change? If we go with this, we can submit this PR first, and deal with shadowing, internal rediects and retries after the change is pushed to nghttp2. But it may take longer. I notice envoy still uses an old version of nghttp2.

soya3129 replied:
To add more details about why the nghttp2 fix ([here](https://github.com/nghttp2/nghttp2/pull/1312)) is needed in redirect, etc cases, but not needed in regular case. 

In the regular case, when we call encodeHeaders() in onPoolReady(), we will submit the headers by calling sendPendingFrames() before encoding any other frames. Calling sendPendingFrames() will make sure headers (session->ob_syn) are popped for processing first, because that's the only frame available. In redirect case, sendPendingFrames() in encodeHeaders() will be skipped because we are still dispatching (dispatching_ is true [here](https://github.com/envoyproxy/envoy/blob/dc3467adacb8f2b0393583857e709c06d76a2441/source/common/http/http2/codec_impl.cc#L671)). So we accumulate both headers and metadata. When sendPendingFrames is executed at the end of dispatch(), metadata shows up first because it is in session->ob_reg ([here](https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2352)), which is processed before session->ob_syn([here](https://github.com/nghttp2/nghttp2/blob/d93842db3eb481fc6953ec70a81ed2f7863ffb45/lib/nghttp2_session.c#L2360)).

It's easiest, and I think it also makes sense to fix in nghttp2. (sync frame should always process before regular frames?). If it takes too long, I will try to find some hacky ways to fix it in Envoy.

alyssawilk replied:
As said I'm totally fine dealing with this later, I'd just like a tracking TODO :-)

soya3129 replied:
Fixed. Thanks!

alyssawilk commented on test/integration/http2_integration_test.cc:
```diff
@@ -109,6 +109,26 @@ TEST_P(Http2MetadataIntegrationTest, ProxyMultipleMetadataReachSizeLimit) {
   testEnvoyMultipleMetadataReachSizeLimit();
 }
 
+TEST_P(Http2MetadataIntegrationTest, ProxySmallMetadataInRequest) {
```
Also sorry I missed this on prior PRs but you can define the tests inline here as they're only used in http2_integration_test.

I'm trying to fix the thing where we define the code in one place and then run the tests elsewhere :-)

soya3129 replied:
Fixed. Sorry it's a little messy because I also moved response tests.

soya3129 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -779,6 +779,25 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
+    if ((*entry)->end_stream_ && insert_new_metadata_in_headers_ && !empty_data_end_stream_sent_) {
+      Buffer::OwnedImpl empty_data("");
+      ENVOY_STREAM_LOG(trace,
+                       "inserts an empty data frame to carry end_stream instead of headers "
```
Fixed. Thanks!

soya3129 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -795,6 +814,7 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
   }
 
   if (continue_data_entry != decoder_filters_.end()) {
+
```
Oops. Fixed.

soya3129 commented on include/envoy/http/filter.h:
```diff
@@ -396,6 +405,14 @@ class StreamDecoderFilter : public StreamFilterBase {
+   *
+   * @param metadata supplies the decoded metadata.
+   */
+  virtual FilterMetadataStatus decodeMetadata(MetadataMap& metadata_map) PURE;
```
Thanks for the feedback!

Actually, because metadata is generally a hop-by-hop message, there is no need to wait for data or events to happen. All filters should return FilterMetadataStatus::Continue no matter they process metadata or not. As a result, metadata's process is kind of independent. When metadata is received or generated, it goes through all the filters's decodeMetadata() immediately, and then send to the next hop right away if the connection pool is ready. If the connection is not ready, metadata is stored in the router filter and wait for the connection to be ready. No further process of the metadata is needed. In summery, it doesn't interact with the process of other frame types, and is not part of the continueDecoding/commonContinue logic.

Sorry I am not familiar with ext_authz filter. Could you please elaborate on the problems that can be caused by metadata? Thanks!

lizan replied:
If `HEADER, METADATA, DATA` is a possible sequence then I do think blindly returning `FilterMetadataStatus::Continue` will be a problem.

Filters like `ext_authz` (as well as isito mixer filter) does call external service to check if the request can be passed to upstream. It initiates a RPC call (REST or gRPC) to an external service and `StopIteration` until the external service respond (or timeout, fail), so in case of `HEADER, METADATA, DATA` sequence, decodeMetadata has to return `StopIteration` as well during the wait for external service. See [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/http/ext_authz/ext_authz.cc#L75) for what it does for `decodeData` and `decodeTrailers`.

Blindly returning `Continue` will result if `METADATA` arrives before external service respond/timeout, it bypass the result of that and being proxied to upstream.

mattklein123 replied:
Yeah agree this is a problem and I'm not sure the best way of handling this, because very few filters are going to care about metadata. This will definitely require filters that do the stop/continue logic that @lizan talks about to be modified. I hate to say this, but I'm wondering if we should make metadata callbacks opt-in somehow and not a primary callback interface. If a filter does not opt-in to metadata, the HCM would deal with the correct stop/continue logic automatically. Thoughts?

htuch replied:
What about a more general model for L7 events, where filters can opt-in to arbitrary events? The existing subclass/inheritance model works great for a few fixed events, but it seems that as we scale we don't want to force breaking API changes on all filters (plus, many don't care about most HTTP pipeline events beyond headers/data/trailers).

This could be done with dynamic registration of callbacks or handler interface implementations at construction time.

soya3129 replied:
Hi @lizan @alyssawilk and @mattklein123 , since we now have the StopAll status for headers ([PR](https://github.com/envoyproxy/envoy/pull/5954)), I am trying to apply it to the existing filters, starting with ext_authz. (Please let me know if we want metadata to use StopAll first, and then existing filters).

I have two questions:
1. Is there a procedure envoy follows to make production changes? If the change breaks production, is there a way to roll back?
2. The second question is about ext_authz filter. Maybe @lizan knows the answer. We can config ext_authz to send auth check after the full request body is received ([here](https://github.com/envoyproxy/envoy/blob/fa69fad0de6b63a254ce6a9f8164b31163a1ada0/source/extensions/filters/http/ext_authz/ext_authz.cc#L92)). But looks like auth check only needs request headers ([here](https://github.com/envoyproxy/envoy/blob/fa69fad0de6b63a254ce6a9f8164b31163a1ada0/source/extensions/filters/http/ext_authz/ext_authz.cc#L25)). I think it could be a way to buffer body when StopAll option is not available or it could be a feature needed by someone. Do you know if we should keep it with the new StopAlll option?

Thanks!

mattklein123 replied:
I can answer (1) which is we definitely revert PRs if they cause production issues so don't worry about that.

lizan replied:
> But looks like auth check only needs request headers ([here](https://github.com/envoyproxy/envoy/blob/fa69fad0de6b63a254ce6a9f8164b31163a1ada0/source/extensions/filters/http/ext_authz/ext_authz.cc#L25)).

@soya3129 No it needs decodingBuffer(), which is retrieved via callback here: https://github.com/envoyproxy/envoy/blob/075edf802b9ef94bdf26fddc9e775f8bb89df92d/source/extensions/filters/common/ext_authz/check_request_utils.cc#L121

soya3129 replied:
Ah, got it. Will preserve data buffering. Thanks!!

alyssawilk replied:
With Matt's permission I'd lean towards making the base class implementation return Continue, and landing these and tests in a follow up PR.  I did a recent change like that for the rc details and I think it is helpful to allow folks to fix their own filter implementations.

Alternately for this one we might want to keep the base class impl around

Either way I think we may have to move some of the existing Envoy filters to StopAllIteration before they work well with metadata.

soya3129 replied:
Added a todo to remind myself.

@alyssawilk, which filters would you recommend to check StopAllIteration? I can try to make the change. Thanks!

mattklein123 replied:
> With Matt's permission I'd lean towards making the base class implementation return Continue, and landing these and tests in a follow up PR. I did a recent change like that for the rc details and I think it is helpful to allow folks to fix their own filter implementations.

Agreed. I think so few filters are going to care about metadata that we should try to handle this transparently. IMO having the base class return continue until we figure out a full solution makes sense to me.

soya3129 replied:
Added default return. Thanks!

mattklein123 commented:
@alyssawilk @soya3129 sorry for the silence. I need to spend some in-depth time with this PR. I've seen my name mentioned a few times, but other than the @lizan comment which I agree is a problem, I think it's all in outdated comments. Can you let me know which are the major open design issues on this PR and I can look at them in the next couple of days? Or should I just do an in-depth pass?

alyssawilk commented:
@mattklein123 I think my high level questions are how we can best avoid interface churn for filter implementers (for this and future functions) and ideas on how to reduce HCM complexity

For outstanding design it is known but not yet TODOd that this doesn't handle retries/redirects but launch and iterate, right?

alyssawilk commented:
The other way we could work around this is implement the NoStopReally filter return code for the other objects which we talked about in some now-stalebotted-out issue.   If we did that, it'd stop all further HCM processing including onData calls that the user doesn't want and onMetadata calls (and I think would be generally useful) and metadata could continue being continue-only.

mattklein123 commented:
> The other way we could work around this is implement the NoStopReally filter return code for the other objects which we talked about in some now-stalebotted-out issue. If we did that, it'd stop all further HCM processing including onData calls that the user doesn't want and onMetadata calls (and I think would be generally useful) and metadata could continue being continue-only.

Yeah agreed this would also work, though it would require all effected filters to be modified to prevent breakage. I think either way we are going to need to think through some HCM changes here?

lizan commented:
> The other way we could work around this is implement the NoStopReally filter return code for the other objects which we talked about in some now-stalebotted-out issue.   If we did that, it'd stop all further HCM processing including onData calls that the user doesn't want and onMetadata calls (and I think would be generally useful) and metadata could continue being continue-only.

How about a filter status like `Passthrough` (suggest any better naming please), that doesn't imply `continue` but just inherit previous return code? i.e. if one `decode*` returned `StopIteration` and subsequent `decode*` returned `Passthrough` it doesn't continue decoding until explicit `continueDecoding` is called or subsequent `decode*` returns `Continue`. In that case it is safe to provide default implementation returning `Passthrough`.

mattklein123 commented:
> How about a filter status like Passthrough (suggest any better naming please), that doesn't imply continue but just inherit previous return code? i.e. if one decode* returned StopIteration and subsequent decode* returned Passthrough it doesn't continue decoding until explicit continueDecoding is called or subsequent decode* returns Continue. In that case it is safe to provide default implementation returning Passthrough.

Yes this would work also. I think we can add this to our list of potential solutions. Either way I think we have to step back and think through this more carefully. @lizan I have a meeting set up with @alyssawilk  on Monday to chat about this if you want to join (through I think the early time is going to be a non-starter for you :wink:).

soya3129 commented:
keep alive.

soya3129 commented:
keep alive.

soya3129 commented:
keep alive.

soya3129 commented:
keep alive.

soya3129 commented:
Reopen the metadata change in the request path. What's new compared to the previous version:

1. Save metadata in saved_request_metadata_ when decodeHeaders() returns StopAllIteration. Resume metadata parsing when iteration continues.
2. Note that, decodeHeaders() may generate metadata. In this case, we don't know decodeHeaders's return status before handling metadata. The solution is to save metadata in saved_request_metadata_ until we know the return status.
3. There is a un-addressed comment from the last review round: how to handle metadata in retry, shadowing and redirect. The plan is to save metadata sent by downstream in ActiveStream, and pass it to router filter when those cases happen. The change should be straightforward. I don't include the change in this PR because we are still waiting a nghttp2 fix (https://github.com/nghttp2/nghttp2/pull/1312) to be pulled before we can handle the recreating stream cases.

mattklein123 commented:
I think @alyssawilk is going to take a pass on this but I'm assuming not till Monday at the earliest.

alyssawilk commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -347,6 +348,11 @@ MetadataDecoder& ConnectionImpl::StreamImpl::getMetadataDecoder() {
 void ConnectionImpl::StreamImpl::onMetadataDecoded(MetadataMapPtr&& metadata_map_ptr) {
+  if (decoder_destroyed_) {
+    // nghttp2 returns error, and ConnectionManager will call resetAllStreams(). Avoid referring to
+    // decoder_.
```
If decoder_ is unsafe to refer to, should it be nulled out rather than being tracked as a separate boolean?  This seems fragile.

soya3129 replied:
Fixed. Thanks!

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -462,6 +462,24 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::HeaderMap& trailers) {
+  ASSERT(upstream_requests_.size() == 1);
+  if (upstream_requests_.front() == nullptr) {
+    ENVOY_STREAM_LOG(trace,
+                     "the upstream_request_ not ready. Storing metadata_map to encode later: {}",
```
upstream_request_ -> upstream_request_ is

alyssawilk replied:
Though unless metadata can race ahead of headers, I don't think we need to worry about this.  I think we want to handle "no upstream connection" in one place.  Right now when we get headers we always create the upstream request, and then eventually when we have a stream to the backend we have an encoder.  I think it'd be good to eagerly create, and then buffer in the upstream request consistently.

soya3129 replied:
I think metadata may arrive before headers when decodeHeaders() inserts new metadata. That because addDecodedMetadata() will call decodeMetadata() through all filters. At that time, headers may only half way through the filters.

alyssawilk replied:
OK, in that case what if we create the upstream request when we receive the first frame (data or metadata) and always let the upstream request buffer?

soya3129 replied:
Good idea! Fixed.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -1198,8 +1245,13 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
+
+  bool end_stream = !buffered_request_body_ && encode_complete_ && !encode_trailers_;
+  // If end_stream is set in headers, and there are metadata to send, delays end_stream. The case
+  // only happens when decoding headers filters return ContinueAndEndStream.
```
Ah, I think this answers my question up above on how we handle the headers only condition.  I think we should probably handle it in the HCM so that other filters don't have to all one-off this.

soya3129 replied:
There are two cases where an empty data frame is needed: 1. the request is headers only request. 2. a filter decides to ignore data and trailers even if the request has data or trailers, aka the decoding_headers_only_  case. Case 1 can be easily handled in HCM. But case 2 is not easy to deal with in HCM. None of the methods I came up with is as simple as the current solution. I added tests for both cases in the latest version.

The purpose of decoding_headers_only_ filter is to skip passing data to the following filters. But if a filter inserts metadata in decodeHeaders() before or after decoding_headers_only_ filter, to deal with it in HCM, we need to pass the empty data frame through the following filters, which conflicts what decoding_headers_only_ filter is doing. So we need to add conditions to make sure only empty frames pass through the following filters, not the data from request, and it only passes through once. Another problem I encountered is which filter the empty data frame should start iterating from. Sometimes, decodeHeaders() returns stop in a filter after decoding_headers_only_ filter. If the continue_data_entry([here](https://github.com/envoyproxy/envoy/blob/5867e33124deac887c7100f9c3b3480e1fd67f42/source/common/http/conn_manager_impl.cc#L837)) is set to decoding_headers_only_ filter (it should be, because that's when we know if we need the empty data frame or not), commonContinue() will call the decodeHeaders in the stopped filter again without reset the iteration state. Since we don't need to pass the empty frame through filters, the safest way is to iterate the empty frame from the router filter (the last one), which I think may be equivalent to the current solution. I am open to any ideas or suggestions.

alyssawilk replied:
I think it's possible this comment is confusing me.

If any intermediate filter added metadata, end stream would not have been set along with the headers by the time the content got to the router filter (or I think we have a problem).  So if just-before-router filter added metadata, the router filter would get headers with end stream false.

The way I see it what this code is doing is determining if this is a headers only request to send to upstream.  If there encode is complete but there's body, trailers, or metadata, it's not a headers only stream.

If it is the case that ContinueAndEndStream results in headers-with-end-stream-true going through the filter chain, wouldn't every filter up to the router filter be assuming it's done before it gets metadata?

soya3129 replied:
When the request is a header-only request, added metadata will unset end stream in headers, and router filter receives headers without end stream set. But if it's a ContinueAndEndStream case, and decodeHeaders() added metadata, I keep the end stream in headers to tell the following filters that request's data and trailers should not be parsed. In this case, metadata are guaranteed to arrive before headers at the router filter. Because addDecodedMetadata() calls decodeMetadata() in every filters before return. And decodeMetadata() can only return continue. The router filter will always get metadata before headers if decodeHeaders() adds metadata. Does that make sense?

If we unset end stream in ContinueAndEndStream case, I think it would be hard/messy to let the following filters skip parsing data and trailers.

alyssawilk replied:
"The router filter will always get metadata before headers if decodeHeaders() adds metadata."

Is this still true with the decode_headers_called_ check in the HCM?  I thought that made sure that metadata never races before headers?

soya3129 replied:
Ah, you are right! With decode_headers_called_, metadata should always arrive after headers! decode_headers_called_ was created to handle a very rare case where a filter inserts new metadata, and another filter follows it returns StopAllIteration. To prevent metadata passing through a filter where the filter's header callback hasn't been called yet, decode_headers_called_ is used to indicate if header callback has been called. I only thought its use in StopAll cases, and didn't thought it has the impact of delay metadata for all statuses. Good catch!

It would simplify router filter's code then. We don't need to consider upstream request creation in decodeMetadata()!

But back to the main point in this comment thread, I think it's still possible that router filter receives headers with end stream set, and then it receives metadata because end_stream is used to tell the following filters that data and trailers should not be parsed.

danzh2010 commented on source/docs/h2_metadata.md:
```diff
@@ -171,3 +169,37 @@ ConnectionManagerImpl::ActiveStream::encodeMetadata(ActiveStreamEncoderFilter\*
+StreamDecoderFilterCallbacks::addDecodedMetadata(). After calling each filter's decoding function,
+Envoy checks if new metadata is added to ActiveStream::request\_metadata\_map\_vector\_. If so,
+then Envoy calls ConnectionManagerImpl::ActiveStream::decodeMetadata(ActiveStreamEncoderFilter\* filter,
+MetadataMapPtr&& metadata\_map) to go through all the filters.
```
Should the newly added metadata go through ALL the filters or the current one and following filters? This PR seems to iterate through all the filters on the new metadata. But previous filters may already regards itself done with the stream (end_stream_ == true). So it's a bit surprising to me to know that the metadata added by decodeData|Trailers  should go through all the filters. I maybe missed some important point, if that's the case, please excuse my ignorance.

soya3129 replied:
Good question! You are right, metadata should start iterating from the current and the following filters. It was a decision made a few PRs back: (please refer to https://github.com/envoyproxy/envoy/pull/5165 for details and search for "current_encoder_filter_" for the discussion around this). About the end_stream concern, as long as end_stream is sent after metadata, we are OK, :) This is handled in router filter.

danzh2010 commented on test/integration/filters/metadata_stop_all_filter.cc:
```diff
@@ -0,0 +1,74 @@
+
+namespace Envoy {
+
+class MetadataStopAllFilter : public Http::PassThroughFilter {
```
This filter does a lot of similar thing as DecodeHeadersReturnStopAllFilter. Can we extend that one and re-use it instead?

soya3129 replied:
I think separating them may make it easier to read the test. DecodeHeadersReturnStopAllFilter is already messy with many special code to verify data StopAll functions. May I keep the current file?

soya3129 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -813,6 +813,20 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
+      Buffer::OwnedImpl empty_data("");
+      ENVOY_STREAM_LOG(
+          trace, "inserting an empty data frame for end_stream due metadata being added.", *this);
+      addDecodedData(*((*entry).get()), empty_data, true);
```
Please see the comments below.

mattklein123 replied:
Presumably we need the empty data frame because we can't set end_stream on metadata, is that right? Might be worth a comment if that is true.

soya3129 replied:
Correct. Comments added.

soya3129 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -813,6 +813,20 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
+      empty_data_end_stream_sent_ = true;
+    }
+
+    (*entry)->decode_headers_called_ = true;
```
Good idea! empty_data_end_stream_sent_ removed. But I didn't merge the metadata to this case. Please see the explanation below.

soya3129 commented on source/common/http/conn_manager_impl.h:
```diff
@@ -369,6 +402,8 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     void encodeTrailers(ActiveStreamEncoderFilter* filter, HeaderMap& trailers);
     void encodeMetadata(ActiveStreamEncoderFilter* filter, MetadataMapPtr&& metadata_map_ptr);
     void maybeEndEncode(bool end_stream);
+    // Returns true if new metadata is inserted. Otherwise, returns false.
```
Fixed. Thanks!

soya3129 commented on include/envoy/http/filter.h:
```diff
@@ -283,6 +283,15 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
+   *
+   * @return a reference to metadata map vector, where new metadata map can be added.
+   */
+  virtual MetadataMapVector& addDecodedMetadata() PURE;
```
It is to be consistent with addDecodedData() and addDecodedTrailers(). But, I agree, it is a little less straightforward, :)

alyssawilk commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -348,12 +348,9 @@ MetadataDecoder& ConnectionImpl::StreamImpl::getMetadataDecoder() {
-    // nghttp2 returns error, and ConnectionManager will call resetAllStreams(). Avoid referring to
-    // decoder_.
-    return;
+  if (decoder_ != nullptr) {
```
I guess what I really don't understand is why metadata is a special case, where after failure we'll still interact with the decoder, where we can get away with asserts on the other paths.

I'd think that either after failure, the state machine should assure that we won't keep trying to interact with the decoder, or there are no higher up guarantees and we need to do checks on all the various decoder access, or metadata really is a special case here and we should comment why :-)

soya3129 replied:
Good point!! I think metadata is not a special case. When a nghttp2 callback function fails, resetAllStream() will be called ([here](https://github.com/envoyproxy/envoy/blob/5e039a6581cf5fa1f7b70329afeb2e3d35b8189c/source/common/http/conn_manager_impl.cc#L284)). After that, ActiveStream (aka, decoder_) will be scheduled to delete. If a client keeps pushing data and before the connection is closed, code_->dispatch() will keep processing data, which may refer to decoder_.

Looks like only three callbacks can return failure: ConnectionImpl::onInvalidFrame(), ConnectionImpl::onMetadataReceived() and ConnectionImpl::onMetadataFrameComplete(). I null out decoder_ in the three functions, and check decoder_ before referring it.

Will check if response direction has the same issue.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -470,6 +476,40 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::HeaderMap& trailers) {
+Http::FilterMetadataStatus Filter::decodeMetadata(Http::MetadataMap& metadata_map) {
+  Http::MetadataMap metadata;
+  metadata.insert(metadata_map.begin(), metadata_map.end());
+  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata);
```
So just to sanity check, we don't need to do the copy for headers because the HCM permanently stores request headers, but we copy metadata because right now the HCM doesn't store metadata, yes?

If so this will get fixed when we handle metadata retries and I'm good with both not creating a copy utility and Ok with the temporary perf hit.  Might be worth a comment that the copy will go away as we continue with metadata functionality.

soya3129 replied:
That's correct! I changed this copy to move since router is the last filter and no one will refer to metadata_map afterwards. In the future, the move will happen in HCM.

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -470,6 +476,40 @@ Http::FilterTrailersStatus Filter::decodeTrailers(Http::HeaderMap& trailers) {
+  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata);
+
+  if (upstream_requests_.front() == nullptr) {
+    ENVOY_STREAM_LOG(trace,
```
My goal in eagerly creating the upstream request was to not have to queue the metadata in two places.  Can we handle route/pool failure here as well, so we don't have to buffer?

soya3129 replied:
Removed upstream_requests_ creation because headers should always arrive before metadata, :D Please see the explanation in the comments below.

soya3129 commented on source/common/http/conn_manager_impl.h:
```diff
@@ -370,6 +403,8 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     void encodeTrailers(ActiveStreamEncoderFilter* filter, HeaderMap& trailers);
     void encodeMetadata(ActiveStreamEncoderFilter* filter, MetadataMapPtr&& metadata_map_ptr);
     void maybeEndEncode(bool end_stream);
+    // Returns true if new metadata decoded. Otherwise, returns false.
```
Fixed. Thanks!

soya3129 commented on source/common/http/conn_manager_impl.h:
```diff
@@ -202,13 +212,25 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
+      }
+      saved_request_metadata_.clear();
+    }
+    // Calls this function after the filter calls decodeHeaders() to drain accumulated metadata.
```
Fixed. Thanks!

soya3129 commented on source/common/http/http2/codec_impl.cc:
```diff
@@ -347,7 +349,9 @@ MetadataDecoder& ConnectionImpl::StreamImpl::getMetadataDecoder() {
 
 void ConnectionImpl::StreamImpl::onMetadataDecoded(MetadataMapPtr&& metadata_map_ptr) {
-  decoder_->decodeMetadata(std::move(metadata_map_ptr));
+  if (decoder_ != nullptr) {
```
Created #6842. Will merge to this PR then #6842 is submitted.

mattklein123 replied:
Per comments in the other PR I think we are changing direction here, right?

alyssawilk commented on source/common/router/router.cc:
```diff
@@ -1130,12 +1151,33 @@ void Filter::UpstreamRequest::encodeTrailers(const Http::HeaderMap& trailers) {
     ENVOY_STREAM_LOG(trace, "buffering trailers", *parent_.callbacks_);
   } else {
+    // Encodes metadata if exists.
+    if (!parent_.downstream_metadata_map_vector_.empty()) {
```
Oops - I'd missed that we were using the same map location for buffered metadata. 
I think we'll want this to be per-upstream for when hedging (#6228) lands.

At that point we'll create a copy of metadata for each  upstream and each upstream will keep its own queue.

soya3129 replied:
Good catch! Moved metadata storage to upstream.

alyssawilk commented:
Oh also you'll want a master merge to fix up the release build OOM

mattklein123 commented:
> I'd welcome a first code pass from @mattklein123 and I'll take one last look at the tests as well.

Will spend time on this tomorrow morning.

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1047,13 +1066,49 @@ void ConnectionManagerImpl::ActiveStream::decodeTrailers(ActiveStreamDecoderFilt
 
+void ConnectionManagerImpl::ActiveStream::decodeMetadata(MetadataMapPtr&& metadata_map) {
+  resetIdleTimer();
+  // After going through filters, the ownership of metadata_map will be passed to router filter.
```
nit: I would probably not talk about router filter in this code as it is not explicitly required (e.g., in the case of another terminal filter).

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1047,13 +1066,49 @@ void ConnectionManagerImpl::ActiveStream::decodeTrailers(ActiveStreamDecoderFilt
+    if (!(*entry)->decode_headers_called_ || (*entry)->stoppedAll()) {
+      Http::MetadataMap metadata;
+      metadata.insert(metadata_map.begin(), metadata_map.end());
+      Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata);
```
I think this is doing a copy? Can you just create a new map in the pointer and then insert into it, avoiding the local variable? More generally, and I'm probably forgetting context that I used to have (sorry), but why don't we move a metadata map in from the codec here so we don't need to copy at all?

soya3129 replied:
Fixed by removing the copy. I am not sure I follow the second question. But I think the new version is equivalent to a move which means no copy?

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1832,6 +1905,19 @@ Buffer::WatermarkBufferPtr ConnectionManagerImpl::ActiveStreamDecoderFilter::cre
 
+void ConnectionManagerImpl::ActiveStreamDecoderFilter::handleMetadataAfterHeadersCallback() {
+  // If we drain accumulated metadata, the iteration must start with the current filter.
+  bool saved_state = iterate_from_current_filter_;
```
nit: const

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -1246,6 +1293,18 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
   } else {
+    // Encode metadata after headers and before any other frame type.
+    if (!downstream_metadata_map_vector_.empty()) {
+      ENVOY_STREAM_LOG(trace, "Send metadata onPoolReady. {}", *parent_.callbacks_,
```
I might bump this to debug but nbd if you feel trace is better.

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -1234,8 +1276,13 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
+  bool end_stream = !buffered_request_body_ && encode_complete_ && !encode_trailers_;
+  // If end_stream is set in headers, and there are metadata to send, delays end_stream. The case
+  // only happens when decoding headers filters return ContinueAndEndStream.
+  bool delay_headers_end_stream = end_stream && !downstream_metadata_map_vector_.empty();
```
nit: const

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -1130,12 +1151,33 @@ void Filter::UpstreamRequest::encodeTrailers(const Http::HeaderMap& trailers) {
+    downstream_metadata_map_vector_.emplace_back(std::move(metadata_map_ptr));
+  } else {
+    ENVOY_STREAM_LOG(trace, "Encode metadata: {}", *parent_.callbacks_, *metadata_map_ptr);
+    Http::MetadataMapVector metadata_map_vector;
```
Similar question about why we need to create a local vector just to encode the map?

soya3129 replied:
This is to convert from unique_ptr<MetadataMap> to MetadataMapVector (aka vector<unique_ptr<MetadataMap>>) to preserve the possible multiple copies of the same metadata. (encodeMetadata's input is MetadataMapVector).

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -1113,6 +1126,14 @@ void Filter::UpstreamRequest::encodeData(Buffer::Instance& data, bool end_stream
 
     buffered_request_body_->move(data);
   } else {
+    // Encodes metadata if exists.
```
When does this happen here and below? In the case where we don't have an encoder, won't we deal with any pending metadata when the pool is ready? Can we instead just assert that we don't have any pending metadata here and below?

soya3129 replied:
Good catch! Removed and added asserts.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -392,8 +395,11 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 
-  UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(*this, *conn_pool);
-  upstream_request->moveIntoList(std::move(upstream_request), upstream_requests_);
+  if (upstream_requests_.front() == nullptr) {
```
This change is confusing to me. Don't we need to make an upstream request when we get headers? I'm trying to reconcile with `decodeMetadata` which seems to indicate that metadata must come after headers.

soya3129 replied:
Oops! Sorry, this change was made when I tried to allow metadata to create upstream request when they arrive before headers. But it turned out metadata are always behind headers, and I forgot to revert this change.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -353,11 +353,14 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
-  if (!conn_pool) {
-    sendNoHealthyUpstreamResponse();
-    return Http::FilterHeadersStatus::StopIteration;
+  Http::ConnectionPool::Instance* conn_pool = nullptr;
```
Can you explain this change more? Related to my comment below.

soya3129 replied:
Removed. Sorry!

mattklein123 replied:
more to revert here?

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on include/envoy/http/filter.h:
```diff
@@ -480,6 +489,16 @@ class StreamDecoderFilter : public StreamFilterBase {
+   * @param metadata supplies the decoded metadata.
+   */
+  virtual FilterMetadataStatus decodeMetadata(MetadataMap& /* metadata_map */) {
+    return Http::FilterMetadataStatus::Continue;
```
Can you add a TODO/substantial comment here about how this default will break filters that do stop and continue style processing and that any use of metadata needs to be carefully audited while we sort out how to handle this generically? (Or something to this effect)

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -832,6 +832,21 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
     ENVOY_STREAM_LOG(trace, "decode headers called: filter={} status={}", *this,
                      static_cast<const void*>((*entry).get()), static_cast<uint64_t>(status));
 
+    bool new_metadata_added = processNewlyAddedMetadata();
```
nit: const

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/extensions/filters/http/common/pass_through_filter.h:
```diff
@@ -18,10 +18,12 @@ class PassThroughDecoderFilter : public virtual StreamDecoderFilter {
   Http::FilterTrailersStatus decodeTrailers(Http::HeaderMap&) override {
     return Http::FilterTrailersStatus::Continue;
   }
+  Http::FilterMetadataStatus decodeMetadata(Http::MetadataMap&) override {
```
Revert all these changes in the filters now that you have a default?

soya3129 replied:
Fixed. Thanks!

soya3129 commented:
I added a TODO in case this change lands before the codec change, since this PR only has test coverage for metadata crash but not the potential invalid frame crash. The invalid frame test is not as easy to create as I thought. We can't use headers as invalid frame since we need headers to create the stream which is referred after delete, and it's hard to use other frame type because they either are created by nghttp2 or close the stream which we don't want since we need to keep pushing data on the same stream. The current plan is to mock client connection so we can corrupt the frame on wire. I am still working on it.

soya3129 commented:
Crash fix and the corresponding test reverted. Will be submitted in a separate PR. Thanks!

soya3129 commented:
This PR is ready for another look. Thanks!

soya3129 commented:
This PR is ready for another look. I will be OOO between Jun13-Jun21. Response may be delayed during that time. Thanks!

mattklein123 commented:
@alyssawilk is it your turn or mine? :)

alyssawilk commented:
I think yours :-)

On Tue, Jun 11, 2019 at 11:06 PM Matt Klein <notifications@github.com>
wrote:

> @alyssawilk <https://github.com/alyssawilk> is it your turn or mine? :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/5656?email_source=notifications&email_token=AELALPONUAKE7ISSP2SOKKTP2BR3LA5CNFSM4GRBY2T2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXPDR6I#issuecomment-501102841>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AELALPLAOIHR32VDRZQJFV3P2BR3LANCNFSM4GRBY2TQ>
> .
>

mattklein123 commented:
ACK. Will review. Sorry for the delay @soya3129.

mattklein123 commented:
@soya3129 sorry for the huge delay. Now that you are back from vacation can you merge master and I will review once done?

/wait

mattklein123 commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1065,13 +1086,47 @@ void ConnectionManagerImpl::ActiveStream::decodeTrailers(ActiveStreamDecoderFilt
+
+  for (; entry != decoder_filters_.end(); entry++) {
+    // If the filter pointed by entry has stopped for all frame type, stores metadata and return.
+    // If the filter pointed by entry hasn't returned from decodeHeaders, stores newly added
```
Sorry when does this happen? Does this somehow happen if inside the headers callback the filter adds metadata? Can you clarify that if so in the comment?

soya3129 replied:
That's right. Added comments to clarify.

mattklein123 commented on source/common/http/conn_manager_impl.h:
```diff
@@ -203,13 +227,25 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
                          ActiveStream::FilterIterationStartState::CanStartFromCurrent);
     }
+    void doMetadata() override { drainSavedRequestMetadata(); }
+    MetadataMapVector& saved_metadata() override { return *getSavedRequestMetadata(); }
```
nit: `savedMetadata()`

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/common/http/conn_manager_impl.h:
```diff
@@ -203,13 +227,25 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     const HeaderMapPtr& trailers() override { return parent_.request_trailers_; }
 
+    void drainSavedRequestMetadata() {
+      for (auto& metadata_map : *getSavedRequestMetadata()) {
```
Perhaps ASSERT that the storage vector pointer is not nullptr so we don't end up creating an empty vector anyway? Can you audit other calls to make sure we aren't accidentally creating a vector just to check that it's empty?

soya3129 replied:
Good point! Fixed in both request and response direction.

alyssawilk replied:
Should we have a parallel assert in drainSavedResponseMetadata?

soya3129 replied:
Fixed. Thanks!

mattklein123 commented on source/common/http/conn_manager_impl.h:
```diff
@@ -521,6 +577,12 @@ class ConnectionManagerImpl : Logger::Loggable<Logger::Id::http>,
     absl::optional<Router::RouteConstSharedPtr> cached_route_;
     absl::optional<Upstream::ClusterInfoConstSharedPtr> cached_cluster_info_;
     std::list<DownstreamWatermarkCallbacks*> watermark_callbacks_{};
+    // Stores metadata added in the decoding filter that is being processed. Will be cleared before
```
I notice that we have similar storage on a per-filter basis as well as on a connection basis. Is it possible to collapse that somehow? If not can you add some clarifying comments on why we need both types of storage? It's not immediately clear.

soya3129 replied:
Added comments and removed not used variables.

mattklein123 commented on source/common/router/router.cc:
```diff
@@ -1477,8 +1501,13 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
 
   upstream_timing_.onFirstUpstreamTxByteSent(parent_.callbacks_->dispatcher().timeSource());
+
+  bool end_stream = !buffered_request_body_ && encode_complete_ && !encode_trailers_;
```
nit: const

soya3129 replied:
Fixed. Thanks!

alyssawilk commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -1065,13 +1086,48 @@ void ConnectionManagerImpl::ActiveStream::decodeTrailers(ActiveStreamDecoderFilt
+      commonDecodePrefix(filter, FilterIterationStartState::CanStartFromCurrent);
+
+  for (; entry != decoder_filters_.end(); entry++) {
+    // If the filter pointed by entry has stopped for all frame type, stores metadata and return.
```
return -> returns

soya3129 replied:
Fixed.

soya3129 commented:
Thanks for the reviews!

mattklein123 commented:
Assuming we cut 1.11.0 next week, I'm tempted to hold this until after that. @soya3129 any objections?

soya3129 commented:
Nope! Thanks for letting me know.
